File: js/src/stream.ts
import { EMPTY_BYTES, concatBytes, toUint8Array, type Bytes } from "./bytes.js";

export interface StreamPackageParserOptions {
  maxBodySize?: number | undefined;
  onPackage?: ((pkg: Uint8Array) => void) | undefined;
}

export function packStreamPackage(pkg: Bytes): Uint8Array {
  const body = toUint8Array(pkg);
  if (body.byteLength > 0xffffffff) {
    throw new RangeError("package length exceeds uint32");
  }

  const packed = new Uint8Array(4 + body.byteLength);
  new DataView(packed.buffer, packed.byteOffset, packed.byteLength).setUint32(0, body.byteLength, true);
  packed.set(body, 4);
  return packed;
}

export class StreamPackageParser {
  private readonly maxBodySize: number;
  private buffer: Uint8Array = EMPTY_BYTES;
  private bodySize: number | undefined;

  onPackage: ((pkg: Uint8Array) => void) | undefined;

  constructor(options: StreamPackageParserOptions = {}) {
    this.maxBodySize = options.maxBodySize ?? 0xffffffff;
    this.onPackage = options.onPackage;
  }

  reset(): void {
    this.buffer = EMPTY_BYTES;
    this.bodySize = undefined;
  }

  feed(chunk: Bytes): void {
    const bytes = toUint8Array(chunk);
    if (bytes.byteLength === 0) {
      return;
    }

    this.buffer = this.buffer.byteLength === 0 ? bytes.slice() : concatBytes([this.buffer, bytes]);

    for (;;) {
      if (this.bodySize === undefined) {
        if (this.buffer.byteLength < 4) {
          return;
        }
        this.bodySize = new DataView(this.buffer.buffer, this.buffer.byteOffset, this.buffer.byteLength).getUint32(0, true);
        if (this.bodySize > this.maxBodySize) {
          this.reset();
          throw new RangeError(`body_size > max_body_size: ${this.bodySize} > ${this.maxBodySize}`);
        }
        this.buffer = this.buffer.slice(4);
      }

      if (this.buffer.byteLength < this.bodySize) {
        return;
      }

      const pkg = this.buffer.slice(0, this.bodySize);
      this.buffer = this.buffer.slice(this.bodySize);
      this.bodySize = undefined;
      this.onPackage?.(pkg);
    }
  }
}

File: python/rpc_core/stream.py
from __future__ import annotations

from typing import Callable


def pack_stream_package(package: bytes | bytearray | memoryview) -> bytes:
    body = bytes(package)
    if len(body) > 0xFFFFFFFF:
        raise ValueError("package length exceeds uint32")
    return len(body).to_bytes(4, "little") + body


class StreamPackageParser:
    def __init__(
        self,
        *,
        max_body_size: int = 0xFFFFFFFF,
        on_package: Callable[[bytes], None] | None = None,
    ) -> None:
        self.max_body_size = max_body_size
        self.on_package = on_package
        self._buffer = b""
        self._body_size: int | None = None

    def reset(self) -> None:
        self._buffer = b""
        self._body_size = None

    def feed(self, chunk: bytes | bytearray | memoryview) -> None:
        data = bytes(chunk)
        if not data:
            return

        self._buffer += data
        while True:
            if self._body_size is None:
                if len(self._buffer) < 4:
                    return
                self._body_size = int.from_bytes(self._buffer[:4], "little")
                if self._body_size > self.max_body_size:
                    size = self._body_size
                    self.reset()
                    raise ValueError(f"body_size > max_body_size: {size} > {self.max_body_size}")
                self._buffer = self._buffer[4:]

            if len(self._buffer) < self._body_size:
                return

            package = self._buffer[: self._body_size]
            self._buffer = self._buffer[self._body_size :]
            self._body_size = None
            if self.on_package is not None:
                self.on_package(package)
